app/src/main/java/io/apicurio/registry/store/RegistryStorage.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 Red Hat
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.registry.store;
+
+import java.util.List;
+
+import io.apicurio.registry.rest.beans.ArtifactMetaData;
+import io.apicurio.registry.rest.beans.EditableMetaData;
+import io.apicurio.registry.rest.beans.Rule;
+import io.apicurio.registry.rest.beans.VersionMetaData;
+
+/**
+ * The storage layer for the registry.
+ * @author eric.wittmann@gmail.com
+ */
+public interface RegistryStorage {
+
+    public ArtifactMetaData createArtifact(String artifactId, String content) throws ArtifactAlreadyExistsException, RegistryStorageException;
+    
+    public void deleteArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public StoredArtifact getArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public ArtifactMetaData updateArtifact(String artifactId, String content) throws ArtifactNotFoundException, RegistryStorageException;
```

Comment:
I would change this operation to `updateArtifactVersion(String artifactId, long version, String content)`. AFAIU the artifact is a sequence of 'artifact versions', so I'm not sure how the update operation would work. Maybe you meant updating the latest version of the artifact? In that case creating a new latest version would work better.

Reply:
Yeah the logic of `updateArtifact` is to create a new version of the artifact, making it the latest.  So this storage method would actually be used (logically) by **both** of these REST operations:

* `PUT` `/artifacts/{artifactId}`
* `POST` `/artifacts/{artifactId}/versions`

My thinking here is that 98.3% of the time clients will be dealing with an artifact as though it is simply a BLOB of content.  They can create, update, and delete that thing.  The ability to look back at older versions is an edge case supported by the API but not optimized.  So even though the back-end will actually be storing every version of the content, the *main focus* of the REST API is the concept of the Artifact itself, not really the collection of Artifact versions (even though those **can be** managed).

Reply:
One other note - version content actually cannot be "updated".  So an Artifact Version is immutable.  The **Artifact** is mutable by creating a new version.